wrappers/services.go

```diff
@@ -375,120 +376,202 @@ func userDaemonReload() error {
+					newFileState.Mode = st.Mode()
+				}
+
+				if mkdirErr := os.MkdirAll(filepath.Dir(path), 0755); mkdirErr != nil {
+					return mkdirErr
+				}
+				ensureErr := osutil.EnsureFileState(path, &newFileState)
+				if ensureErr != nil && ensureErr != osutil.ErrSameState {
+					// fatal error trying to write the file
+					return ensureErr
+				}
+				if ensureErr == nil {
+					// then we modified the file and should save this as something
+					// to roll back to
+					if haveOldFile {
+						st := osutil.FileState(&oldFileState)
+						modifiedUnits[path] = &st
+					} else {
+						modifiedUnits[path] = nil
+					}
+					modified[s] = append(modified[s], app)
+				}
+
+				switch app.DaemonScope {
+				case snap.SystemDaemon:
+					modifiedSystem = true
+				case snap.UserDaemon:
+					modifiedUser = true
+				}
+
+				return nil
+			}
+
+			// create services first; this doesn't trigger systemd
+
+			// Generate new service file state
+			path := app.ServiceFile()
 			var content []byte
-			content, err = generateSnapTimerFile(app)
+			content, err = generateSnapServiceFile(app, opts)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			path := app.Timer.File()
-			os.MkdirAll(filepath.Dir(path), 0755)
-			if err = osutil.AtomicWriteFile(path, content, 0644, 0); err != nil {
-				return err
+
+			err = handleFile(path, content)
+			if err != nil {
+				return nil, err
+			}
+
+			// Generate systemd .socket files if needed
+			var socketFiles *map[string][]byte
+			socketFiles, err = generateSnapSocketFiles(app)
+			if err != nil {
+				return nil, err
+			}
+			for path, content := range *socketFiles {
+				err = handleFile(path, content)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			if app.Timer != nil {
+				content, err = generateSnapTimerFile(app)
+				if err != nil {
+					return nil, err
+				}
+				path := app.Timer.File()
+
+				err = handleFile(path, content)
+				if err != nil {
+					return nil, err
+				}
 			}
-			written = append(written, path)
 		}
 	}
 
 	if !preseeding {
-		if writtenSystem {
+		if modifiedSystem {
 			if err = sysd.DaemonReload(); err != nil {
-				return err
+				return nil, err
 			}
 		}
-		if writtenUser {
+		if modifiedUser {
 			if err = userDaemonReload(); err != nil {
-				return err
+				return nil, err
 			}
 		}
 	}
 
-	return nil
+	return modified, nil
+}
+
+// AddSnapServicesOptions is a struct for controlling the generated service
+// definition for a snap service.
+type AddSnapServicesOptions struct {
+	Preseeding              bool
+	VitalityRank            int
+	RequireMountedSnapdSnap bool
+}
+
+// AddSnapServices adds service units for the applications from the snap which
+// are services. The services do not get enabled or started.
+func AddSnapServices(s *snap.Info, opts *AddSnapServicesOptions, inter interacter) error {
+	m := map[*snap.Info]*AddSnapServicesOptions{
+		s: opts,
+	}
+
+	_, err := EnsureSnapServices(m, inter)
```

Comment:
We don't need to export `EnsureSnapService` currently, do we?

Reply:
no, not currently so I can unexport it if you like, but we will just have to re-export it in the followup PR's so it just seemed like a waste of bits to unexport it only to re-export it in the next PR